File: SimpleCoreData/Modules/Start/ViewController/StartViewController.m
//
//  StartViewController.m
//  SimpleCoreData
//
//  Created by Kasper Spychala on 03/10/2024.
//

#import <Foundation/Foundation.h>

#import "StartViewController.h"
#import "SimpleCoreData-Swift.h"

@interface StartViewController ()

@property (nonatomic, strong) StartView *contentView;

@end

@implementation StartViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    [self handleButton];
    [self updateCompanyName];
}

- (void)loadView {
    [super loadView];

    self.view = [[UIView alloc] initWithFrame:[UIScreen mainScreen].bounds];
    self.view.backgroundColor = [UIColor whiteColor];

    self.contentView = [[StartView alloc] initWithFrame:CGRectZero];
    self.contentView.translatesAutoresizingMaskIntoConstraints = NO;
    [self.view addSubview:self.contentView];

    [NSLayoutConstraint activateConstraints:@[
        [self.contentView.topAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.topAnchor],
        [self.contentView.bottomAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.bottomAnchor],
        [self.contentView.leadingAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.leadingAnchor],
        [self.contentView.trailingAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.trailingAnchor]
    ]];
}

- (void)updateCompanyName {
    NSString *companyName = [self.viewModel getCompanyName];
    if (companyName == nil) {
        NSLog(@"Company name not accessible");
        return;
    }
    [self.contentView updateCompanyNameWithName:companyName];
}

- (void)handleButton {
    __weak typeof(self) weakSelf = self;
    self.contentView.onButtonTap = ^{
        __strong typeof(weakSelf) strongSelf = weakSelf;
        [strongSelf.viewModel navigateToSubmit];
    };
}

@end

File: SimpleCoreData/Modules/Start/ViewController/StartViewController.swift
//
//  StartViewController.swift
//  SimpleCoreData
//
//  Created by Kasper Spychala on 02/10/2024.
//

import UIKit

class StartViewController: UIViewController {
    var viewModel: StartViewModelProtocol?
    private let contentView: StartView = {
        let view = StartView(frame: .zero)
        view.translatesAutoresizingMaskIntoConstraints = false
        return view
    }()

    override func viewDidLoad() {
        super.viewDidLoad()
        handleButton()
        updateCompanyName()
    }

    private func updateCompanyName() {
        guard let companyName = viewModel?.getCompanyName() else {
            LoggingManager.shared.error("Company name not accessible")
            return
        }
        contentView.updateCompanyName(name: companyName)
    }

    private func handleButton() {
        contentView.onButtonTap = { [weak self] in
            guard let self = self else { return }
            self.viewModel?.navigateToSubmit()
        }
    }

    override func loadView() {
        super.loadView()
        self.view = UIView()
        self.view.backgroundColor = .white
        self.view.addSubview(contentView)
        NSLayoutConstraint.activate([
            contentView.topAnchor.constraint(
                equalTo: self.view.safeAreaLayoutGuide.topAnchor
            ),
            contentView.bottomAnchor.constraint(
                equalTo: self.view.safeAreaLayoutGuide.bottomAnchor
            ),
            contentView.leadingAnchor.constraint(
                equalTo: self.view.safeAreaLayoutGuide.leadingAnchor
            ),
            contentView.trailingAnchor.constraint(
                equalTo: self.view.safeAreaLayoutGuide.trailingAnchor
            )
        ])
    }
}

